### PR TITLE
Remove write barrier exemption for T_ICLASS

### PR DESCRIPTION
**Pull request by XrXr**
Due to T_ICLASS sharing the method table and the constant table with
other objects, adding a method or a constant used to mean creating
multiple edges in the object graph. This made inserting write barriers
(WBs) difficult, so iclasses and modules with usages were marked as
WB unprotected.

 - Instead of sharing IV_TABLE between T_ICLASS and T_MODULE, go
   from T_ICLASS to T_MODULE to access the table. IV_TABLE was used
   for implementing class variable access and autoload.
 - Move the method table and the constant table onto the GC heap.
   Addtions to these table now only create one edge in the object
   reference graph, making it feasible to have these tables WB
   protected.
 - Remove `RICLASS_ORIGIN_SHARED_MTBL` as method tables are now
   always GC owned. The flag was used to do manual ownership
   management.
 - Layout struct rb_id_table to have members that are important for
   performing lookups as close to the beginning as possible. Doing
   this increases the chance that lookups complete in two cacheline
   loads.
 - Insert write barriers for T_ICLASS and bless it as WB protected

Please see [[Feature #16984]](https://bugs.ruby-lang.org/issues/16984) for performance numbers and other details.

### CODE REVIEW

**Comment by XrXr** [comment]
Hmm I'm not sure why this is running into `SIGILL`. I'm looking into it.

**Comment by ko1** [comment]
@tenderlove explained this plan before. Is it same one?

**Comment by XrXr** [comment]
Yes, the idea for this is from @tenderlove.

File: gc.c [hunk 1]
```diff
@@ -2817,12 +2810,6 @@ obj_free(rb_objspace_t *objspace, VALUE obj)
...
       case T_ICLASS:
-	/* Basically , T_ICLASS shares table with the module */
-        if (FL_TEST(obj, RICLASS_IS_ORIGIN) &&
-                !FL_TEST(obj, RICLASS_ORIGIN_SHARED_MTBL)) {
-            /* Method table is not shared for origin iclasses of classes */
-            rb_id_table_free(RCLASS_M_TBL(obj));
-        }
 	if (RCLASS_CALLABLE_M_TBL(obj) != NULL) {
 	    rb_id_table_free(RCLASS_CALLABLE_M_TBL(obj));
 	}
```
Comment (tenderlove) [inline gc.c:2815]: This method table is managed by the GC now, right?  Can we remove this if statement?

Comment (XrXr) [reply gc.c:2815]: `CALLABLE_M_TBL` is still manually managed. It's not shared between objects AFAIK. I'm not sure if there would be any benefits from moving this table onto the GC heap.

File: id_table.c [hunk 2]
```diff
@@ -99,6 +95,21 @@ rb_id_table_create(size_t capa)
     return rb_id_table_init(tbl, (int)capa);
 }
 
+struct rb_managed_id_table *
+rb_new_managed_id_table(int memo_type, size_t capa)
```
Comment (ko1) [inline id_table.c:99]: I prefer `rb_managed_id_table_new()` .

File: id_table.c [hunk 3]
```diff
@@ -99,6 +95,21 @@ rb_id_table_create(size_t capa)
...
+    struct rb_managed_id_table *table = (struct rb_managed_id_table *)rb_imemo_new(memo_type, 0, 0, 0, 0);
+    rb_id_table_init(&table->table, (int)capa);
+    return table;
+}
+
+void
+rb_id_table_deallocate(struct rb_managed_id_table *tbl)
+{
+    xfree(tbl->table.items);
+    MEMZERO(&tbl->table, struct rb_id_table, 1);
```
Comment (ko1) [inline id_table.c:110]: Also `rb_managed_id_table_free()` is preferable for me.

File: id_table.h [hunk 4]
```diff
@@ -15,12 +15,30 @@ enum rb_id_table_iterator_result {
     ID_TABLE_ITERATOR_RESULT_END
 };
 
+typedef struct rb_id_item item_t;
```
Comment (ko1) [inline id_table.h:18]: `item_t` should not be exposed, IMO.

File: id_table.h [hunk 5]
```diff
@@ -15,12 +15,30 @@ enum rb_id_table_iterator_result {
...
+
+/* id_table imemo */
+struct rb_managed_id_table {
+    VALUE flags;
+    struct rb_id_table table;
+};
+
 struct rb_id_table *rb_id_table_create(size_t size);
+struct rb_id_table *rb_id_table_init(struct rb_id_table *tbl, int capa);
+struct rb_managed_id_table *rb_new_managed_id_table(int memo_type, size_t capa);
```
Comment (ko1) [inline id_table.h:34]: Separating `rb_id_table_*` decls and `rb_managed_id_table_*` will be clear for me.

File: internal/class.h [hunk 6]
```diff
@@ -37,9 +37,9 @@ struct rb_classext_struct {
     struct st_table *iv_index_tbl;
     struct st_table *iv_tbl;
 #if SIZEOF_SERIAL_T == SIZEOF_VALUE /* otherwise m_tbl is in struct RClass */
-    struct rb_id_table *m_tbl;
+    struct rb_managed_id_table *m_tbl;
```
Comment (ko1) [inline internal/class.h:40]: (trivial) renaming `m_tbl` will helpful to detect the code which doesn't respect the type (using cast)  and macro.

File: id_table.h (hunk 4 above)
Comment (XrXr) [reply id_table.h]: I moved the typedef to `id_table.c`. Let me know if you prefer having it as a void pointer instead of `struct rb_id_item *`.

**Comment by XrXr** [comment]
Closing in favor of #3410

**Pull request closed without merging**